File: src/utils/formatPredicates.ts
// ── Condition normalizer ──────────────────────────────────────────────────────

function normalizeCondition(raw: string): string {
  return (
    raw
      .trim()
      // Multi-char operators first (must precede single-char passes)
      .replace(/\s*>=\s*/g, ' >= ')
      .replace(/\s*<=\s*/g, ' <= ')
      .replace(/\s*!=\s*/g, ' != ')
      .replace(/\s*==\s*/g, ' == ')
      // Single-char comparisons — only when not part of a two-char operator
      .replace(/([^<>!=])\s*>\s*(?![=])/g, '$1 > ')
      .replace(/([^<>!=])\s*<\s*(?![=])/g, '$1 < ')
      // Logical keywords — collapse surrounding whitespace
      .replace(/\s+and\s+/gi, ' and ')
      .replace(/\s+or\s+/gi, ' or ')
      .replace(/\bnot\s+/gi, 'not ')
      // Collapse multiple spaces
      .replace(/\s{2,}/g, ' ')
      .trim()
  )
}

// ── Action normalizer ─────────────────────────────────────────────────────────

function normalizeSingleAction(act: string): string {
  const t = act.trim()
  if (!t) return ''

  // read(x)
  if (/^read\s*\(/.test(t)) return t.replace(/^read\s*\(/, 'read(')

  // print(x)
  if (/^print\s*\(/.test(t)) return t.replace(/^print\s*\(/, 'print(')

  // goto label
  const gotoMatch = t.match(/^goto\s+(\w+)$/)
  if (gotoMatch) return `goto ${gotoMatch[1]}`

  // assignment: lhs = rhs  (normalize spaces around =)
  const eqIdx = t.indexOf('=')
  if (eqIdx > 0) {
    const lhs = t.slice(0, eqIdx).trim()
    const rhs = t.slice(eqIdx + 1).trim()
    if (/^[A-Za-z_][A-Za-z0-9_]*$/.test(lhs)) return `${lhs} = ${rhs}`
  }

  return t
}

function normalizeActions(raw: string): string {
  const acts = raw.split(';').map(normalizeSingleAction).filter(Boolean)
  return acts.length ? acts.join('; ') + ';' : ''
}

// ── Line formatter ────────────────────────────────────────────────────────────

function formatLine(line: string): string {
  const trimmed = line.trim()
  if (!trimmed) return ''

  let rest = trimmed
  const parts: string[] = []

  // Optional @thread prefix
  const threadMatch = rest.match(/^@(\w+)\s*(.*)$/)
  if (threadMatch) {
    parts.push(`@${threadMatch[1]}`)
    rest = threadMatch[2].trim()
  }

  // Optional :label prefix
  const labelMatch = rest.match(/^:(\w+)\s*(.*)$/)
  if (labelMatch) {
    parts.push(`:${labelMatch[1]}`)
    rest = labelMatch[2].trim()
  }

  // { condition }
  const condMatch = rest.match(/^\{([^}]*)\}\s*(.*)$/)
  if (!condMatch) return trimmed // unparseable — return unchanged
  parts.push(`{ ${normalizeCondition(condMatch[1])} }`)
  rest = condMatch[2].trim()

  // [ actions ]
  const actMatch = rest.match(/^\[([^\]]*)\]\s*(.*)$/)
  if (!actMatch) return trimmed
  parts.push(`[ ${normalizeActions(actMatch[1])} ]`)
  rest = actMatch[2].trim()

  // Optional <sem: resource>
  const memoMatch = rest.match(/<([^:>]+):([^>]+)>/)
  if (memoMatch) parts.push(`<${memoMatch[1].trim()}: ${memoMatch[2].trim()}>`)

  return parts.join(' ')
}

// ── Public API ────────────────────────────────────────────────────────────────

export function formatPredicates(text: string): string {
  const lines = text.split('\n').map(formatLine)

  // Collapse consecutive blank lines into a single blank line
  const out: string[] = []
  let prevBlank = false
  for (const line of lines) {
    const blank = line === ''
    if (blank && prevBlank) continue
    out.push(line)
    prevBlank = blank
  }

  return out.join('\n').trim()
}

File: poc-cpp/format_predicates.cpp
// ─────────────────────────────────────────────────────────────────────────────
// PoC #2: C++ port of src/utils/formatPredicates.ts  (REGEX-HEAVY module)
//
// Purpose: measure the real cost of porting a regex-dense module. Unlike the
// stack parser (char-classes only), this file is ~15 regex replacements incl.
// lookahead (?![=]), negated classes [^<>!=], case-insensitive keywords, and
// the temporal U/X canonicalization.
//
// Finding: std::regex uses the ECMAScript grammar, so nearly every JS regex
// here ports 1:1 — lookahead, \b, \s, char classes, backreferences ($1) all
// work. The ONE JS feature std::regex lacks is *lookbehind* — this module does
// not use it (it uses a negated char class instead), so it ports cleanly.
// (See lookbehind_demo.cpp for the codeGenerator case that DOES need a rewrite.)
//
// Build: clang++ -std=c++17 -O2 format_predicates.cpp -o format_predicates
// ─────────────────────────────────────────────────────────────────────────────
#include <string>
#include <vector>
#include <regex>
#include <iostream>
#include <sstream>

namespace {
using std::string;
using std::regex;

string trim(const string& s) {
  size_t a = s.find_first_not_of(" \t\r\n\f\v");
  if (a == string::npos) return "";
  size_t b = s.find_last_not_of(" \t\r\n\f\v");
  return s.substr(a, b - a + 1);
}

// ── normalizeCondition — the regex-dense core ─────────────────────────────────
string normalizeCondition(const string& raw) {
  string s = trim(raw);
  // Each line mirrors one .replace() in the TS source, same order.
  s = std::regex_replace(s, regex(R"(\s*>=\s*)"), " >= ");
  s = std::regex_replace(s, regex(R"(\s*<=\s*)"), " <= ");
  s = std::regex_replace(s, regex(R"(\s*!=\s*)"), " != ");
  s = std::regex_replace(s, regex(R"(\s*==\s*)"), " == ");
  // single-char comparisons — negated class + lookahead, both supported
  s = std::regex_replace(s, regex(R"(([^<>!=])\s*>\s*(?![=]))"), "$1 > ");
  s = std::regex_replace(s, regex(R"(([^<>!=])\s*<\s*(?![=]))"), "$1 < ");
  // logical keywords — case-insensitive
  s = std::regex_replace(s, regex(R"(\s+and\s+)", std::regex::icase), " and ");
  s = std::regex_replace(s, regex(R"(\s+or\s+)",  std::regex::icase), " or ");
  s = std::regex_replace(s, regex(R"(\bnot\s+)",  std::regex::icase), "not ");
  // temporal operators — 'until' → 'U' (first only), leading 'next' → 'X'
  s = std::regex_replace(s, regex(R"(\s+(?:U|[Uu]ntil)\s+)"), " U ",
                         std::regex_constants::format_first_only);
  s = std::regex_replace(s, regex(R"(^(?:X|[Nn]ext)\b)"), "X");
  // collapse multiple spaces
  s = std::regex_replace(s, regex(R"(\s{2,})"), " ");
  return trim(s);
}

// ── normalizeSingleAction ─────────────────────────────────────────────────────
string normalizeSingleAction(const string& act) {
  string t = trim(act);
  if (t.empty()) return "";

  if (std::regex_search(t, regex(R"(^read\s*\()")))
    return std::regex_replace(t, regex(R"(^read\s*\()"), "read(");
  if (std::regex_search(t, regex(R"(^print\s*\()")))
    return std::regex_replace(t, regex(R"(^print\s*\()"), "print(");

  std::smatch m;
  if (std::regex_match(t, m, regex(R"(^goto\s+(\w+)$)")))
    return "goto " + m[1].str();

  size_t eq = t.find('=');
  if (eq != string::npos && eq > 0) {
    string lhs = trim(t.substr(0, eq));
    string rhs = trim(t.substr(eq + 1));
    if (std::regex_match(lhs, regex(R"(^[A-Za-z_][A-Za-z0-9_]*$)")))
      return lhs + " = " + rhs;
  }
  return t;
}

string normalizeActions(const string& raw) {
  std::vector<string> acts;
  std::stringstream ss(raw);
  string part;
  while (std::getline(ss, part, ';')) {
    string a = normalizeSingleAction(part);
    if (!a.empty()) acts.push_back(a);
  }
  if (acts.empty()) return "";
  string out;
  for (size_t i = 0; i < acts.size(); i++) { if (i) out += "; "; out += acts[i]; }
  return out + ";";
}

// ── formatLine ────────────────────────────────────────────────────────────────
string formatLine(const string& line) {
  string trimmed = trim(line);
  if (trimmed.empty()) return "";

  string rest = trimmed;
  std::vector<string> parts;
  std::smatch m;

  if (std::regex_match(rest, m, regex(R"(^@(\w+)\s*([\s\S]*)$)"))) {
    parts.push_back("@" + m[1].str());
    rest = trim(m[2].str());
  }
  if (std::regex_match(rest, m, regex(R"(^:(\w+)\s*([\s\S]*)$)"))) {
    parts.push_back(":" + m[1].str());
    rest = trim(m[2].str());
  }
  if (!std::regex_match(rest, m, regex(R"(^\{([^}]*)\}\s*([\s\S]*)$)")))
    return trimmed; // unparseable — unchanged
  parts.push_back("{ " + normalizeCondition(m[1].str()) + " }");
  rest = trim(m[2].str());

  if (!std::regex_match(rest, m, regex(R"(^\[([^\]]*)\]\s*([\s\S]*)$)")))
    return trimmed;
  parts.push_back("[ " + normalizeActions(m[1].str()) + " ]");
  rest = trim(m[2].str());

  if (std::regex_search(rest, m, regex(R"(<([^:>]+):([^>]+)>)")))
    parts.push_back("<" + trim(m[1].str()) + ": " + trim(m[2].str()) + ">");

  string out;
  for (size_t i = 0; i < parts.size(); i++) { if (i) out += " "; out += parts[i]; }
  return out;
}

// ── formatPredicates (public) ─────────────────────────────────────────────────
string formatPredicates(const string& text) {
  std::vector<string> lines;
  std::stringstream ss(text);
  string line;
  while (std::getline(ss, line, '\n')) lines.push_back(formatLine(line));

  std::vector<string> out;
  bool prevBlank = false;
  for (const string& l : lines) {
    bool blank = l.empty();
    if (blank && prevBlank) continue;
    out.push_back(l);
    prevBlank = blank;
  }
  string joined;
  for (size_t i = 0; i < out.size(); i++) { if (i) joined += "\n"; joined += out[i]; }
  return trim(joined);
}

} // namespace

int main() {
  std::stringstream buf; buf << std::cin.rdbuf();
  std::cout << formatPredicates(buf.str());
  return 0;
}
